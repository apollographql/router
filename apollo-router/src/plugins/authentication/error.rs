use displaydoc::Display;
use jsonwebtoken::Algorithm;
use jsonwebtoken::errors::Error as JWTError;
use jsonwebtoken::errors::ErrorKind;
use jsonwebtoken::jwk::KeyAlgorithm;
use serde::Deserialize;
use serde::Serialize;
use thiserror::Error;
use tower::BoxError;

#[derive(Debug, Display, Error)]
pub(crate) enum AuthenticationError {
    /// Configured header is not convertible to a string
    CannotConvertToString,

    /// Value of '{0}' JWT header should be prefixed with '{1}'
    InvalidJWTPrefix(String, String),

    /// Value of '{0}' JWT header has only '{1}' prefix but no JWT token
    MissingJWTToken(String, String),

    /// '{0}' is not a valid JWT header: {1}
    InvalidHeader(String, JWTError),

    /// Cannot create decoding key: {0}
    CannotCreateDecodingKey(JWTError),

    /// JWK does not contain an algorithm
    JWKHasNoAlgorithm,

    /// Cannot decode JWT: {0}
    CannotDecodeJWT(JWTError),

    /// Cannot insert claims into context: {0}
    CannotInsertClaimsIntoContext(BoxError),

    /// Cannot find kid: '{0:?}' in JWKS list
    CannotFindKID(String),

    /// Cannot find a suitable key for: alg: '{0:?}', kid: '{1:?}' in JWKS list
    CannotFindSuitableKey(Algorithm, Option<String>),

    /// Invalid issuer: the token's `iss` was '{token}', but signed with a key from JWKS configured to only accept from '{expected}'
    InvalidIssuer { expected: String, token: String },

    /// Invalid audience: the token's `aud` was '{actual}', but '{expected}' was expected
    InvalidAudience { actual: String, expected: String },

    /// Unsupported key algorithm: {0}
    UnsupportedKeyAlgorithm(KeyAlgorithm),
}

fn jwt_error_to_reason(jwt_err: &JWTError) -> &'static str {
    let kind = jwt_err.kind();
    match kind {
        ErrorKind::InvalidToken => "INVALID_TOKEN",
        ErrorKind::InvalidSignature => "INVALID_SIGNATURE",
        ErrorKind::InvalidEcdsaKey => "INVALID_ECDSA_KEY",
        ErrorKind::InvalidRsaKey(_) => "INVALID_RSA_KEY",
        ErrorKind::RsaFailedSigning => "RSA_FAILED_SIGNING",
        ErrorKind::InvalidAlgorithmName => "INVALID_ALGORITHM_NAME",
        ErrorKind::InvalidKeyFormat => "INVALID_KEY_FORMAT",
        ErrorKind::MissingRequiredClaim(_) => "MISSING_REQUIRED_CLAIM",
        ErrorKind::ExpiredSignature => "EXPIRED_SIGNATURE",
        ErrorKind::InvalidIssuer => "INVALID_ISSUER",
        ErrorKind::InvalidAudience => "INVALID_AUDIENCE",
        ErrorKind::InvalidSubject => "INVALID_SUBJECT",
        ErrorKind::ImmatureSignature => "IMMATURE_SIGNATURE",
        ErrorKind::InvalidAlgorithm => "INVALID_ALGORITHM",
        ErrorKind::MissingAlgorithm => "MISSING_ALGORITHM",
        ErrorKind::Base64(_) => "BASE64_ERROR",
        ErrorKind::Json(_) => "JSON_ERROR",
        ErrorKind::Utf8(_) => "UTF8_ERROR",
        ErrorKind::Crypto(_) => "CRYPTO_ERROR",
        // ErrorKind is non-exhaustive
        _ => "UNKNOWN_ERROR",
    }
}

impl AuthenticationError {
    pub(crate) fn as_context_object(&self) -> ErrorContext {
        let (code, reason) = match self {
            AuthenticationError::CannotConvertToString => ("CANNOT_CONVERT_TO_STRING", None),
            AuthenticationError::InvalidJWTPrefix(_, _) => ("INVALID_PREFIX", None),
            AuthenticationError::MissingJWTToken(_, _) => ("MISSING_JWT", None),
            AuthenticationError::InvalidHeader(_, jwt_err) => {
                ("INVALID_HEADER", Some(jwt_error_to_reason(jwt_err).into()))
            }
            AuthenticationError::CannotCreateDecodingKey(jwt_err) => (
                "CANNOT_CREATE_DECODING_KEY",
                Some(jwt_error_to_reason(jwt_err).into()),
            ),
            AuthenticationError::JWKHasNoAlgorithm => ("JWK_HAS_NO_ALGORITHM", None),
            AuthenticationError::CannotDecodeJWT(jwt_err) => (
                "CANNOT_DECODE_JWT",
                Some(jwt_error_to_reason(jwt_err).into()),
            ),
            AuthenticationError::CannotInsertClaimsIntoContext(_) => {
                ("CANNOT_INSERT_CLAIMS_INTO_CONTEXT", None)
            }
            AuthenticationError::CannotFindKID(_) => ("CANNOT_FIND_KID", None),
            AuthenticationError::CannotFindSuitableKey(_, _) => ("CANNOT_FIND_SUITABLE_KEY", None),
            AuthenticationError::InvalidIssuer { .. } => ("INVALID_ISSUER", None),
            AuthenticationError::InvalidAudience { .. } => ("INVALID_AUDIENCE", None),
            AuthenticationError::UnsupportedKeyAlgorithm(_) => ("UNSUPPORTED_KEY_ALGORITHM", None),
        };

        ErrorContext {
            message: self.to_string(),
            code: code.into(),
            reason,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub(super) struct ErrorContext {
    pub(super) message: String,
    pub(super) code: String,
    pub(super) reason: Option<String>,
}

#[derive(Error, Debug)]
pub(crate) enum Error {
    #[error("header_value_prefix must not contain whitespace")]
    BadHeaderValuePrefix,
}
