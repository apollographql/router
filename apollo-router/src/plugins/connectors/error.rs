//! Connectors error types.

use apollo_federation::sources::connect::Connector;

use crate::graphql;
use crate::graphql::ErrorExtension;
use crate::json_ext::Path;

/// Errors that apply to all connector types. These errors represent a problem invoking the
/// connector, as opposed to an error returned from the connector itself.
#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub(crate) enum Error {
    /// Request limit exceeded
    RequestLimitExceeded,

    /// Offline snapshots are in use, but no snapshot was found
    SnapshotNotFound,
}

impl Error {
    /// Create a GraphQL error from this error.
    #[must_use]
    pub(crate) fn to_graphql_error(
        &self,
        connector: &Connector,
        path: Option<Path>,
    ) -> crate::error::Error {
        let builder = graphql::Error::builder()
            .message(self.to_string())
            .extension_code(self.extension_code())
            .extension("service", connector.id.label.clone());
        if let Some(path) = path {
            builder.path(path).build()
        } else {
            builder.build()
        }
    }
}

impl ErrorExtension for Error {
    fn extension_code(&self) -> String {
        match self {
            Self::RequestLimitExceeded => "REQUEST_LIMIT_EXCEEDED",
            Self::SnapshotNotFound => "SNAPSHOT_NOT_FOUND",
        }
        .to_string()
    }
}
