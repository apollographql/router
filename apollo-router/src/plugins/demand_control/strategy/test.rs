use apollo_compiler::ExecutableDocument;

use crate::plugins::demand_control::strategy::StrategyImpl;
use crate::plugins::demand_control::test::TestError;
use crate::plugins::demand_control::test::TestStage;
use crate::plugins::demand_control::CostContext;
use crate::plugins::demand_control::DemandControlError;
use crate::services::execution::Request;
use crate::services::subgraph::Response;

/// Test strategy for demand control.
/// Can be configured to fail at different stages of the request processing.
pub(crate) struct Test {
    pub(crate) stage: TestStage,
    pub(crate) error: TestError,
}

impl StrategyImpl for Test {
    fn on_execution_request(&self, request: &Request) -> Result<(), DemandControlError> {
        request.context.extensions().with_lock(|mut lock| {
            let cost_context = lock.get_or_default_mut::<CostContext>();
            match self {
                Test {
                    stage: TestStage::ExecutionRequest,
                    error,
                } => Err(cost_context.result(error.into())),
                _ => Ok(()),
            }
        })
    }

    fn on_subgraph_request(
        &self,
        request: &crate::services::subgraph::Request,
    ) -> Result<(), DemandControlError> {
        request.context.extensions().with_lock(|mut lock| {
            let cost_context = lock.get_or_default_mut::<CostContext>();
            match self {
                Test {
                    stage: TestStage::SubgraphRequest,
                    error,
                } => Err(cost_context.result(error.into())),
                _ => Ok(()),
            }
        })
    }

    fn on_subgraph_response(
        &self,
        _request: &ExecutableDocument,
        response: &Response,
    ) -> Result<(), DemandControlError> {
        response.context.extensions().with_lock(|mut lock| {
            let cost_context = lock.get_or_default_mut::<CostContext>();
            match self {
                Test {
                    stage: TestStage::SubgraphResponse,
                    error,
                } => Err(cost_context.result(error.into())),
                _ => Ok(()),
            }
        })
    }

    fn on_execution_response(
        &self,
        context: &crate::Context,
        _request: &ExecutableDocument,
        _response: &crate::graphql::Response,
    ) -> Result<(), DemandControlError> {
        context.extensions().with_lock(|mut lock| {
            let cost_context = lock.get_or_default_mut::<CostContext>();
            match self {
                Test {
                    stage: TestStage::ExecutionResponse,
                    error,
                } => Err(cost_context.result(error.into())),
                _ => Ok(()),
            }
        })
    }
}
