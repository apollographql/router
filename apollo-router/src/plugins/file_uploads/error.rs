use bytesize::ByteSize;
use thiserror::Error;

use crate::graphql;

/// Errors that may occur during file upload
#[derive(Debug, Error)]
pub(super) enum FileUploadError {
    /// Represents an invalid request, wrapping the context as a string
    #[error("invalid multipart request: {0}")]
    InvalidMultipartRequest(#[from] multer::Error),

    #[error("Missing multipart field 'operations', it should be a first field in request body.")]
    MissingOperationsField,

    #[error("Missing multipart field 'map', it should be a second field in request body.")]
    MissingMapField,

    #[error("Invalid JSON in the ‘map’ multipart field: {0}")]
    InvalidJsonInMapField(serde_json::Error),

    #[error("Batched requests are not supported for file uploads.")]
    BatchRequestAreNotSupported,

    #[error("Invalid path '{0}' found inside 'map' field, it should start with 'variables.'.")]
    InvalidPathInsideMapField(String),

    #[error("Invalid path '{0}' found inside 'map' field, missing name of variable.")]
    MissingVariableNameInsideMapField(String),

    #[error("Invalid path '{0}' found inside 'map' field, it does not point to a valid value inside 'operations' field.")]
    InputValueNotFound(String),

    #[error("Missing files in the request: {0}.")]
    MissingFiles(String),

    #[error("Variables containing files are forbidden inside @defer: {0}.")]
    VariablesForbiddenInsideDefer(String),

    #[error("Variables containing files are forbidden inside subscription: {0}.")]
    VariablesForbiddenInsideSubscription(String),

    #[error("References to variables containing files are ordered in the way that prevent streaming of files.")]
    MisorderedVariables,

    #[error("Variables use mutiple time in the way that prevent streaming of files: {0}.")]
    DuplicateVariableUsages(String),

    #[error("Exceeded the limit of {0} file uploads of files in a single request.")]
    MaxFilesLimitExceeded(usize),

    #[error("Exceeded the limit of {limit} on {filename} file.")]
    MaxFileSizeLimitExceeded { limit: ByteSize, filename: String },

    #[error("{0}")]
    HyperBodyErrorWrapper(#[from] hyper::Error),
}

impl From<FileUploadError> for graphql::Error {
    fn from(value: FileUploadError) -> Self {
        Self::builder()
            .message(value.to_string())
            .extension_code(match &value {
                FileUploadError::MaxFilesLimitExceeded(_) => {
                    "FILE_UPLOADS_LIMITS_MAX_FILES_EXCEEDED".to_string()
                }
                FileUploadError::MaxFileSizeLimitExceeded { .. } => {
                    "FILE_UPLOADS_LIMITS_MAX_FILE_SIZE_EXCEEDED".to_string()
                }
                _ => "FILE_UPLOADS_OPERATION_CANNOT_STREAM".to_string(),
            })
            .build()
    }
}
