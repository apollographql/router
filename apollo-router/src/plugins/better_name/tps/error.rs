//! Error types

use std::error;
use std::fmt;

use crate::graphql;

/// The rate limit error.
#[derive(Debug, Default)]
pub(crate) struct TpsLimited;

impl TpsLimited {
    /// Construct a new TpsLimited error
    pub(crate) fn new() -> Self {
        TpsLimited {}
    }
}

impl fmt::Display for TpsLimited {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // TODO: product-y words
        f.pad("TODO: product-specified words")
    }
}

impl From<TpsLimited> for graphql::Error {
    fn from(_: TpsLimited) -> Self {
        graphql::Error::builder()
            .message(String::from("TODO: product words"))
            // TODO: decide on extension code; product insight
            .extension_code("TPS_RATE_LIMITED")
            .build()
    }
}

impl error::Error for TpsLimited {}
