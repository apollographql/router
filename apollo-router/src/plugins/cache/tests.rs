use std::collections::HashMap;
use std::sync::Arc;

use bytes::Bytes;
use fred::error::RedisErrorKind;
use fred::mocks::MockCommand;
use fred::mocks::Mocks;
use fred::prelude::RedisError;
use fred::prelude::RedisValue;
use parking_lot::Mutex;
use tower::ServiceExt;

use super::entity::EntityCache;
use crate::cache::redis::RedisCacheStorage;
use crate::plugin::test::MockSubgraph;
use crate::services::supergraph;
use crate::Context;
use crate::MockedSubgraphs;
use crate::TestHarness;

const SCHEMA: &str = r#"schema
        @core(feature: "https://specs.apollo.dev/core/v0.1")
        @core(feature: "https://specs.apollo.dev/join/v0.1")
        @core(feature: "https://specs.apollo.dev/inaccessible/v0.1")
         {
        query: Query
        subscription: Subscription
   }
   directive @core(feature: String!) repeatable on SCHEMA
   directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet) on FIELD_DEFINITION
   directive @join__type(graph: join__Graph!, key: join__FieldSet) repeatable on OBJECT | INTERFACE
   directive @join__owner(graph: join__Graph!) on OBJECT | INTERFACE
   directive @join__graph(name: String!, url: String!) on ENUM_VALUE
   directive @inaccessible on OBJECT | FIELD_DEFINITION | INTERFACE | UNION
   scalar join__FieldSet
   enum join__Graph {
       USER @join__graph(name: "user", url: "http://localhost:4001/graphql")
       ORGA @join__graph(name: "orga", url: "http://localhost:4002/graphql")
   }
   type Query {
       currentUser: User @join__field(graph: USER)
   }

   type Subscription @join__type(graph: USER) {
        userWasCreated: User
   }

   type User
   @join__owner(graph: USER)
   @join__type(graph: ORGA, key: "id")
   @join__type(graph: USER, key: "id"){
       id: ID!
       name: String
       activeOrganization: Organization
   }
   type Organization
   @join__owner(graph: ORGA)
   @join__type(graph: ORGA, key: "id")
   @join__type(graph: USER, key: "id") {
       id: ID
       creatorUser: User
       name: String
       nonNullId: ID!
       suborga: [Organization]
   }"#;

#[derive(Debug)]
pub(crate) struct Mock1 {
    set: Mutex<bool>,
}

impl Mock1 {
    fn new() -> Mock1 {
        Mock1 {
            set: Mutex::new(false),
        }
    }
}

static USER_RESPONSE:&str = "{\"control\":{\"created\":1705069368},\"data\":{\"currentUser\":{\"activeOrganization\":{\"__typename\":\"Organization\",\"id\":\"1\"}}}}";
static ORGA_RESPONSE:&str = "{\"control\":{\"created\":1705072093},\"data\":{\"creatorUser\":{\"__typename\":\"User\",\"id\":2}}}";
impl Mocks for Mock1 {
    fn process_command(&self, command: MockCommand) -> Result<RedisValue, RedisError> {
        println!("received redis command: {command:?}");

        match &*command.cmd {
            "GET" => {
                if let Some(RedisValue::Bytes(b)) = command.args.get(0) {
                    if b == &b"subgraph:user:Query:146a735f805c55554b5233253c17756deaa6ffd06696fafa4d6e3186e6efe592:d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c"[..]{
                        let set = self.set.lock();
                        if *set {
                            return Ok(RedisValue::Bytes(Bytes::from(USER_RESPONSE)));
                        }
                    } else if b == &b"subgraph:orga:Organization:5811967f540d300d249ab30ae681359a7815fdb5d3dc71a94be1d491006a6b27:655f22a6af21d7ffe671d3ce4b33464a76ddfea0bf179740b15e804b11983c04:d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c"[..] {
                        return Ok(RedisValue::Bytes(Bytes::from(ORGA_RESPONSE)));
                    }
                }
            }
            "SET" => {
                if let Some(RedisValue::Bytes(b)) = command.args.get(0) {
                    if b ==
                        &b"subgraph:user:Query:146a735f805c55554b5233253c17756deaa6ffd06696fafa4d6e3186e6efe592:d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c"[..] {
                            let mut set = self.set.lock();
                            *set = true;

                            //FIXME: can't assert because the creatin date changes
                            //assert_eq!(USER_RESPONSE, command.args.get(1).unwrap().as_str().unwrap(), );
                            return Ok(RedisValue::Null)
                    }
                }
            }
            _ => {}
        }
        Err(RedisError::new(RedisErrorKind::NotFound, "mock not found"))
    }
}

#[tokio::test]
async fn insert() {
    let query = "query { currentUser { activeOrganization { id creatorUser { __typename id } } } }";

    let subgraphs = MockedSubgraphs([
        ("user", MockSubgraph::builder().with_json(
                serde_json::json!{{"query":"{currentUser{activeOrganization{__typename id}}}"}},
                serde_json::json!{{"data": {"currentUser": { "activeOrganization": {
                    "__typename": "Organization",
                    "id": "1"
                } }}}}
            ).build()),
        ("orga", MockSubgraph::builder().with_json(
            serde_json::json!{{
                "query": "query($representations:[_Any!]!){_entities(representations:$representations){...on Organization{creatorUser{__typename id}}}}",
            "variables": {
                "representations": [
                    {
                        "id": "1",
                        "__typename": "Organization",
                    }
                ]
            }}},
            serde_json::json!{{"data": {
                "_entities": [{
                    "creatorUser": {
                        "__typename": "User",
                        "id": 2
                    }
                }]
            }}}
        ).build())
    ].into_iter().collect());

    let redis_cache = RedisCacheStorage::from_mocks(Arc::new(Mock1::new()))
        .await
        .unwrap();
    let entity_cache = EntityCache::with_mocks(redis_cache.clone(), HashMap::new())
        .await
        .unwrap();

    let service = TestHarness::builder()
        .configuration_json(serde_json::json!({"include_subgraph_errors": { "all": true } }))
        .unwrap()
        .schema(SCHEMA)
        .extra_plugin(entity_cache)
        .extra_plugin(subgraphs)
        .build_supergraph()
        .await
        .unwrap();

    let request = supergraph::Request::fake_builder()
        .query(query)
        .context(Context::new())
        .build()
        .unwrap();
    let response = service
        .oneshot(request)
        .await
        .unwrap()
        .next_response()
        .await
        .unwrap();

    insta::assert_json_snapshot!(response);

    // Now testing without any mock subgraphs, all the data should come from the cache
    let entity_cache = EntityCache::with_mocks(redis_cache.clone(), HashMap::new())
        .await
        .unwrap();

    let service = TestHarness::builder()
        .configuration_json(serde_json::json!({"include_subgraph_errors": { "all": true } }))
        .unwrap()
        .schema(SCHEMA)
        .extra_plugin(entity_cache)
        .build_supergraph()
        .await
        .unwrap();

    let request = supergraph::Request::fake_builder()
        .query(query)
        .context(Context::new())
        .build()
        .unwrap();
    let response = service
        .oneshot(request)
        .await
        .unwrap()
        .next_response()
        .await
        .unwrap();

    insta::assert_json_snapshot!(response);
}
