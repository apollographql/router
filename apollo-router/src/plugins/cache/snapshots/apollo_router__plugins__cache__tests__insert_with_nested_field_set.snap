---
source: apollo-router/src/plugins/cache/tests.rs
expression: cache_keys
---
[
  {
    "key": "version:1.0:subgraph:products:type:Query:hash:6173063a04125ecfdaf77111980dc68921dded7813208fdf1d7d38dfbb959627:data:d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c",
    "status": "new",
    "cache_control": "public"
  },
  {
    "key": "version:1.0:subgraph:users:type:User:entity:210e26346d676046faa9fb55d459273a43e5b5397a1a056f179a3521dc5643aa:representation:7cd02a08f4ea96f0affa123d5d3f56abca20e6014e060fe5594d210c00f64b27:hash:2820563c632c1ab498e06030084acf95c97e62afba71a3d4b7c5e81a11cb4d13:data:d9d84a3c7ffc27b0190a671212f3740e5b8478e84e23825830e97822e25cf05c",
    "status": "new",
    "cache_control": "public"
  }
]
